src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs

```diff
@@ -669,6 +669,15 @@ public void WebApplicationBuilder_EnablesServiceScopeValidationByDefaultInDevelo
             Assert.ThrowsAny<Exception>(() => builder.Build());
         }
 
+        [Fact]
+        public async Task WebApplicationBuilder_ThrowsExceptionIfServicesAlreadyBuilt()
+        {
+            var builder = WebApplication.CreateBuilder();
+            await using var app = builder.Build();
+
+            Assert.Throws<InvalidOperationException>(() => builder.Services.AddSingleton<IServer>(new MockAddressesServer()));
```

Comment:
This test needs more asserts (one for each type of mutation).

Reply:
Added a few more assertions here. I think it gives us enough coverage.

FWIW: All of these map to the `TryAdd` extension method which map to `IServiceCollection.Add` internally so it all converges to the same codepath.

Reply:
You need a Remove and an indexer write.